### ISSUE TITLE
Gather operator's output violates the protocol

### ISSUE THREAD

**Issue by FedyuninV** [onnx/onnx#3640]
# Bug Report

### Is the issue related to model conversion?
No. The model was obtained by exporting from another framework but it looks like a protocol mistake. Because `onnx.shape_inference`, `onnxruntime` and `pytorch.export` implement Gather in the same manner. The problem is that this 'manner' doesn't correlate with the protocol.

### Describe the bug
The protocol for [Gather-11](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Gather-11) says:

> axis = 1 :
> 
> Let k = indices[i_{0}, ..., i_{q-1}] Then output[i_{0}, ..., i_{q-1}, j_{0}, ..., j_{r-2}] = input[j_{0}, k, j_{1}, ..., j_{r-2}]

In model I have the following gather  (op name: `Gather_38`):
- `data`: `tensor(float32)` of shape `[1,2,11,1]`
- `indices`: `tensor(int)` of shape `[9,3]`
- `axis`: `2`

The example above shows that first `rank-of-indices` dims are derived from `indices` and next `rank-of-data - 1` dims are derived from `data`.
That's why I expect output shape to be equal to `[9, 3, 1, 2, 1]`. But got `[1, 2, 9, 3, 1]` (similar to `tf.gather`).

### System information
- OS Platform and Distribution (*e.g. Linux Ubuntu 16.04*): Win10
- ONNX version (*e.g. 1.7*):  1.10.1
- Python version: 3.7.9

### Reproduction instructions
- Describe the code to reproduce the behavior.
```
import onnx
from onnx import shape_inference
model = onnx.load('torch-unfold-11.onnx')
inferred_model = shape_inference.infer_shapes(model)
inferred_model.graph.value_info[38] # name:47 is the Gather_38 output
```
- Attach the ONNX model to the issue (where applicable)
https://abbyyihq-my.sharepoint.com/:u:/g/personal/valery_fedyunin_abbyy_com/EaXdV_jcw7NIqA5ReA1CxJgBb35abxIU_SmTWdiIYrn59Q?e=WYWCxM

### Expected behavior
Tensor of shape `[9, 3, 1, 2, 1]`

### Notes

It looks like just an issue with the examples in the protocol. No code fix required.
The only question I have: is it _actually supposed to be of shape_ `[1, 2, 9, 3, 1]` in your protocol?

**Comment by gramalingam** [comment]
I think the documentation is correct. Please note that the attribute "axis" affects the result. The documentation illustrates what happens for axis=0 and axis=1. In your example, axis=2, which is why the result is different.

**Comment by FedyuninV** [comment]
@gramalingam 
I disagree here.

The documentation for `axis=1` shows that first `q` dimensions of the output will be equal to dimensions of the indices:
`output[i_{0}, ..., i_{q-1}, j_{0}, ..., j_{r-2}]`.

And that leads to the output shape to be `indices_shape[0], ..., indices_shape[q-1], data_shape[0], data_shape[2], ..., data_shape[r-1]`.

From this example I make a conclusion: output shape will **always** start with the shape of the indices, because in both cases `axis=0` and `axis=1` it starts with indices shape ([link with both examples](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Gather-11)).

But **actually** the output shape is formed from the data shape, with `axis`'th dimension replaced with indices shape. And **correct** example for `axis=1` must say: `output[j_{0}, i_{0}, i_{1}, ..., i_{q-1}, j_{2}, ..., j_{r-1}]`. That's why I don't get expected output.

**Comment by gramalingam** [comment]
@FedyuninV , you are right. I checked only the right-hand-side, where ```k``` shifts to the specified axis. But the left-hand-side is inconsistent with the shape inferred.
